File: geekbang/algorithm/source-code/3/lc589 N叉树的前序遍历.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        seq = new ArrayList<Integer>();
        if (root == null) return seq;
        Stack<Node> stack = new Stack<Node>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            seq.add(node.val);
            for (int i = node.children.size() - 1; i >= 0; i--) {
                stack.push(node.children.get(i));
            }
        }
        return seq;
    }

    List<Integer> seq;
}

File: geekbang/algorithm/source-code/3/lc589 N叉树的前序遍历.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        q = collections.deque([])
        seq = []
        if not root:
            return seq
        q.append([root, 0])
        while len(q) > 0:
            # Don't use list.pop(0), which is O(N)
            # deque.popleft() is O(1)
            node, depth = q.popleft()
            if depth >= len(seq):
                seq.append([])
            seq[depth].append(node.val)
            for child in node.children:
                q.append([child, depth + 1])
        return seq